merkle/merkle_path.go

```diff
@@ -138,6 +154,113 @@ func pathFromNodeToRootAtSnapshot(node int64, level int, snapshot int64, maxBitL
+		return fetches, fmt.Errorf("recomputePastSnapshot: %d does not exist for tree of size %d", snapshot, treeSize)
+	}
+
+	// We're recomputing the right hand path, the one to the last leaf
+	level := 0
+	// This is the index of the last node in the snapshot
+	lastNode := snapshot - 1
+	// This is the index of the last node that actually exists in the underlying tree
+	lastNodeAtLevel := treeSize - 1
+
+	// Work up towards, the root we may find the node we need without needing to rehash if
+	// it turns out that the left siblings and parents exist all the way up to the level we're
+	// recalculating.
+	for (lastNode & 1) != 0 {
+		if nodeLevel == level {
+			// Then we want a copy of the node at this level
+			glog.V(vvLevel).Infof("copying l:%d ln:%d", level, lastNode)
+			nodeID, err := siblingIDSkipLevels(snapshot, lastNodeAtLevel, level, lastNode^1, maxBitlen)
+			if err != nil {
+				return []NodeFetch{}, err
+			}
+
+			glog.V(vvLevel).Infof("copy node at %s", nodeID.CoordString())
+			return append(fetches, NodeFetch{Rehash: false, NodeID: nodeID}), nil
+		}
+
+		// Left sibling and parent exist at this snapshot and don't need to be rehashed
+		glog.V(vvLevel).Infof("move up ln:%d level:%d", lastNode, level)
+		lastNode >>= 1
+		lastNodeAtLevel >>= 1
+		level++
+	}
+
+	glog.V(vvLevel).Infof("done ln:%d level:%d", lastNode, level)
+
+	// lastNode is now the index of a left sibling with no right sibling. This is where the
+	// rehashing starts
+	savedNodeID, err := siblingIDSkipLevels(snapshot, lastNodeAtLevel, level, lastNode^1, maxBitlen)
+	glog.V(vvLevel).Infof("root for recompute is: %s", savedNodeID.CoordString())
+	if err != nil {
+		return []NodeFetch{}, err
+	}
+
+	if nodeLevel == level {
+		glog.V(vvLevel).Info("emit root (1)")
+		return append(fetches, NodeFetch{Rehash: true, NodeID: savedNodeID}), nil
+	}
+
+	rehash := false
+	subRootEmitted := false  // whether we've added the recomputed subtree root to the path yet
+
+	for lastNode != 0 {
+		glog.V(vvLevel).Infof("in loop level:%d ln:%d lnal:%d", level, lastNode, lastNodeAtLevel)
+
+		if (lastNode & 1) != 0 {
+			nodeID, err := siblingIDSkipLevels(snapshot, lastNodeAtLevel, level, (lastNode-1)^1, maxBitlen)
+			if err != nil {
+				return []NodeFetch{}, err
+			}
+
+			if !rehash && !subRootEmitted {
+				glog.V(vvLevel).Info("emit root (2)")
+				fetches = append(fetches, NodeFetch{Rehash: true, NodeID: savedNodeID})
+				subRootEmitted = true
+			}
+
+			glog.V(vvLevel).Infof("rehash with %s", nodeID.CoordString())
+			fetches = append(fetches, NodeFetch{Rehash: true, NodeID: nodeID})
+			rehash = true
+		}
+
+		lastNode >>= 1
+		lastNodeAtLevel >>= 1
+		level++
+
+		if nodeLevel == level && !subRootEmitted {
+			glog.V(vvLevel).Info("emit root (3)")
+			return append(fetches, NodeFetch{Rehash: rehash, NodeID: savedNodeID}), nil
+		}
+
+		// Exit early if we've gone far enough up the tree to hit the level we're recomputing
+		if level == nodeLevel {
```

Comment:
It doesn't really help IMO because the different exit cases require different work.